### PR TITLE
Fix #2151, Fix #2083: Sponsored background images

### LINKED ISSUE #2083: Incorrect setting toggle for NTP images
### Description: 
Incorrect setting toggle for NTP images

### Steps to Reproduce 
https://bravesoftware.slack.com/archives/C06UXF3KJ/p1575551486274100

**Actual result:** 
https://bravesoftware.slack.com/archives/C06UXF3KJ/p1575551502274300?thread_ts=1575551486.274100&cid=C06UXF3KJ

**Expected result:**
https://bravesoftware.slack.com/archives/C06UXF3KJ/p1575551512275100?thread_ts=1575551486.274100&cid=C06UXF3KJ

**Reproduces how often:** [Easily reproduced, Intermittent Issue]
Easy

**Brave Version:** 
1.14.1(19.12.05.00)

**Device details:** 
All 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### LINKED ISSUE #2151: NTP Branded Backgrounds
### Description: 

Branded NTP should be supported. This includes a lot of smaller tweaks as well:

- [x] Add image credit to the collection view so it does not cover favorites.
- [x] Landscape redesign
- [x] Remove background long-press
- [x] Remove sponsored background tap
- [x] Disable vertical bouncing
- [x] Landscape centering

Branded behavior:
- [x] Re-enabling (or disabling) background images should automatically re-enable (or disable) branded images.
- [x] When disabling Background Images in settings, all favorites and frequently visited sites are displayed as per pre-NTP background image.
- [x] No NTP branded images to be shown in private mode.
- [x] Fix pre-defined rotation of branded images (1 in 4)
- [x] Show branded image as second image
- [x] Pre-defined rotation of branded images
- [x] Brand overlay
- [x] Clicking brand → website
- [x] All rewards popups

### PR DESCRIPTION
**Pull request by jhreis**
## Summary of Changes

This pull request fixes issues #2151 & #2083

## Submitter Checklist:

- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Reviewer Checklist:

- [x] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [x] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [x] Adequate unit test coverage exists to prevent regressions.
- [x] Adequate test plan exists for QA to validate (if applicable).
- [x] Issue is assigned to a milestone (should happen at merge time).

### CODE REVIEW

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 1]
```diff
@@ -269,14 +277,14 @@ class FavoritesViewController: UIViewController, Themeable {
         
         let adsAvailableInRegion = BraveAds.isCurrentLocaleSupported()
         
-        // Don't show notifications if background images are disabled.
-        guard let bgImageInfo = backgroundImage.info else { return nil }
+        if !Preferences.NewTabPage.backgroundImages.value { return nil }
         
+        let isSponsoredImage = background?.sponsor != nil
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift:298]: @jhreis pls confirm if this is enough to determine image sponsorship

Comment (jhreis) [reply Client/Frontend/Browser/HomePanel/FavoritesViewController.swift:298]: It is enough to determine if the _current_ image is a sponsored image.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 2]
```diff
@@ -956,7 +959,8 @@ class BrowserViewController: UIViewController {
         if favoritesViewController == nil {
             let homePanelController = FavoritesViewController(profile: profile,
                                                               fromOverlay: !inline,
-                                                              rewards: rewards)
+                                                              rewards: rewards,
+                                                              background: backgroundDataSource.newBackground())
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController.swift]: We might want to have a mechanism to reload background image when favorites VC is still active.
For example when privacy mode changes, kind of related to https://github.com/brave/brave-ios/issues/2229

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 3]
```diff
@@ -490,6 +508,18 @@ class FavoritesViewController: UIViewController, Themeable {
             make.left.equalTo(left)
             make.top.bottom.equalTo(self.view)
         }
+        
+        imageSponsorButton.snp.remakeConstraints {
+            let parent = imageSponsorButton.superview!
+            $0.height.width.equalTo(170)
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: Can use `.size` when both height and width are equal

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 4]
```diff
@@ -490,6 +508,18 @@ class FavoritesViewController: UIViewController, Themeable {
             make.left.equalTo(left)
             make.top.bottom.equalTo(self.view)
         }
+        
+        imageSponsorButton.snp.remakeConstraints {
+            let parent = imageSponsorButton.superview!
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: Unwrap and provide safe fallback mechanism

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 5]
```diff
@@ -498,18 +528,18 @@ class FavoritesViewController: UIViewController, Themeable {
...
-        guard let info = backgroundImage.info, let name = info.credit?.name else { return }
+        guard let name = background?.wallpaper.credit?.name else { return }
         
-        hideImageCredit = info.isSponsored
+        hideImageCredit = background?.sponsor != nil
         let photoByText = String(format: Strings.photoBy, name)
         imageCreditInternalButton.setTitle(photoByText, for: .normal)
     }
     
+    // TODO: combine with reset?
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: This still applicable?

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 6]
```diff
@@ -544,24 +575,25 @@ class FavoritesViewController: UIViewController, Themeable {
             // the right side cannot drop under `width` (or superview's right side), otherwise whitespace will be shown on right.
             $0.right.greaterThanOrEqualToSuperview()
             
+            // Same as left / right above but necessary for landscape y centering (to prevent overflow
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: Didn't close bracket :D

File: Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+        /// Only available for normal wallpapers, not for sponsored images
+        let credit: Credit?
+        
+        struct Credit: Decodable {
+            let name: String
+            let url: String?
+        }
+        
+        // TODO: Make optional
+        struct FocalPoint: Decodable {
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift]: Does this work well with optional focal points already?

File: Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+        
+        /// Only available for normal wallpapers, not for sponsored images
+        let credit: Credit?
+        
+        struct Credit: Decodable {
+            let name: String
+            let url: String?
+        }
+        
+        // TODO: Make optional
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift]: Still need to do this?

File: Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    ///     e.g. if `0` it should show the `0th` sponsored image _when a sponsored image should be shown_
+    /// This does _not_ impact _when_ a sponsored image is shown, but rather _what_ sponsored image is shown.
+    /// The maximum of this value is undetermined, as that depends on how many images exist in the current sponsorship.
+    // This can be easily converted to a preference to persist
+    private var sponsoredBackgroundRotationIndex = 0
+
+    private lazy var sponsor: Sponsor? = {
+        let sponsoredFilePath = "ntp-sponsored"
+        guard let sponsoredData = self.loadData(file: sponsoredFilePath) else { return nil }
+        // TODO: Wrap
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift]: remove if fixed

File: Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift [hunk 10]
```diff
@@ -0,0 +1,178 @@
...
+        do {
+            let sponsor = try JSONDecoder().decode(Sponsor.self, from: sponsoredData)
+            return sponsor.wallpapers.isEmpty == true ? nil : sponsor
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift]: `== true` part can be removed

File: Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift [hunk 11]
```diff
@@ -0,0 +1,178 @@
...
+        } catch {
+            // Only set a sponsor if there are valid backgrounds
+            return nil
+        }
+    }()
+    
+    private lazy var standardBackgrounds: [Background] = {
+        let backgroundFilePath = "ntp-data"
+        guard let backgroundData = self.loadData(file: backgroundFilePath) else { return [] }
+        // TODO: Wrap
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift]: fix or remove

File: Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift [hunk 12]
```diff
@@ -0,0 +1,178 @@
...
+        }()
+        
+        // Force back to `0` if at end
+        backgroundRotationCounter %= NewTabPageBackgroundDataSource.sponsorshipShowRate
+        // Increment regardless, this is a counter, not an index, so smallest should be `1`
+        backgroundRotationCounter += 1
+        
+        // Item is returned based on our special index.
+        
+        return (backgroundSet[backgroundIndex], useSponsor ? sponsor : nil)
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift:174]: Can use `backgroundSet[safe: backgroundIndex]` so it never crashes when out of bounds

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift [hunk 13]
```diff
@@ -686,6 +720,6 @@ extension FavoritesViewController: FavoriteCellDelegate {
 
 extension FavoritesViewController: PreferencesObserver {
     func preferencesDidChange(for key: String) {
-        self.resetBackground()
+        self.resetBackgroundImage()
```
Comment (iccub) [inline Client/Frontend/Browser/HomePanel/FavoritesViewController.swift:722]: This changes for every preference in the app right?

**Review by iccub — CHANGES_REQUESTED** [review]
Mostly cleanup comments.
I ran the code and it works very well.

File: Client/Frontend/Browser/BrowserViewController.swift (hunk 2 above)
Comment (jhreis) [reply Client/Frontend/Browser/BrowserViewController.swift]: It does this now.

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift (hunk 5 above)
Comment (jhreis) [reply Client/Frontend/Browser/HomePanel/FavoritesViewController.swift]: Nope, fixed now.

File: Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift (hunk 12 above)
Comment (jhreis) [reply Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift:174]: That returns an optional though. There should be no reason for this to crash, as it is chosen from the range, but we could build in more safeguards?

File: Client/Frontend/Browser/HomePanel/FavoritesViewController.swift (hunk 13 above)
Comment (jhreis) [reply Client/Frontend/Browser/HomePanel/FavoritesViewController.swift:722]: No only:
```
Preferences.NewTabPage.backgroundImages.observe(from: self)
Preferences.NewTabPage.backgroundSponsoredImages.observe(from: self)
```
https://github.com/brave/brave-ios/pull/2203/files/9887873dbed1788a0ffc7c3f5f3d159d54d9dd73#diff-9dddcde1fc3411fdf044fc8d4e5a79d0R241

File: Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift (hunk 12 above)
Comment (iccub) [reply Client/Frontend/Browser/HomePanel/NewTabPageBackgroundDataSource.swift:174]: fine by me

**Review by iccub — APPROVED** [review]
Good job, ran the code for a bit and all seems to be working

**Pull request merged**